include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+    }
+
+    if (from.size() != to.size()) {
+      CRAB_ERROR(domain_name(), "::rename different lengths");
+    }
+
+    // update var_map and rev_var_map
+    // extract base_variable_vector for from and to
+    base_variable_vector_t old_base_vars, new_base_vars;
+    for (unsigned i = 0; i < from.size(); ++i) {
+      variable_t old_var = from[i];
+      auto search = m_var_map.find(old_var);
+      if (search == m_var_map.end()) { // NOT FOUND
+        continue;
+      }
+      ghost_variables_t old_ghost_vars = search->second;
+      // append base_var, [offset, size] into old_base_vars
+      old_ghost_vars.add(old_base_vars);
+      // remove from rev_var_map
+      old_ghost_vars.remove_rev_varmap(m_rev_var_map);
+      // remove from var_map
+      m_var_map.erase(old_var);
+
+      variable_t new_var = to[i];
+      // create/get new ghost variable for new_var
+      const ghost_variables_t &new_gvars = get_or_insert_gvars(new_var);
+      // append base_var, [offset, size] into new_base_vars
+      new_gvars.add(new_base_vars);
+    }
+
+    // rename mem domain
+    m_mem.rename(old_base_vars, new_base_vars);
+    // rename cache_lines domain
+    m_cache_lines.rename(old_base_vars, new_base_vars);
+    // rename regs domain
+    m_regs.rename(old_base_vars, new_base_vars);
+    // rename addrs domain
+    m_addrs.rename(old_base_vars, new_base_vars);
+
+    // rename the rest of environments
+    m_rgn_counting_dom.rename(from, to);
+    m_rgn_init_dom.rename(from, to);
+    if (!crab_domain_params_man::get().region_skip_unknown_regions()) {
+      m_rgn_type_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_allocation_sites()) {
+      m_alloc_site_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_deallocation()) {
+      m_rgn_dealloc_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_tag_analysis()) {
+      m_tag_env.rename(from, to);
+    }
+  }
+
+  /* begin intrinsics operations */
+  void intrinsic(std::string name, const variable_or_constant_vector_t &inputs,
+                 const variable_vector_t &outputs) override {
+    //=================================================================//
+    //       Special intrinsics supported by the region domain
+    //=================================================================//
+    // ---DSA region analysis---
+    //      This analysis indicates which regions might belong to the
+    //      same memory object. The intrinstic is added only if object
+    //      has more than one field.
+    // TODO: need to support other analysis
+    auto error_if_not_rgn = [&name](const variable_t &var) {
+      if (!var.get_type().is_region()) {
+        CRAB_ERROR("Intrinsics ", name, " parameter ", var,
+                   " should be a region");
+      }
+    };
+
+    if (is_bottom()) {
+      return;
+    }
+
+    if (name == "regions_from_memory_object") {
+      // why pass to the outputs vector?
```

Comment:
It's not. That line calls the method `intrinsic` from a basic block (https://github.com/seahorn/crab/blob/master/include/crab/cfg/cfg.hpp#L2903), not from the abstract domain. That will be called later. 

```
const statement_t *intrinsic(std::string name,
                               const std::vector<variable_t> &lhs,
                               const std::vector<variable_or_constant_t> &args) {
    return insert(new intrinsic_t(name, lhs, args, this));
  }
 ```